### ISSUE TITLE
Allowing multiple-workers read with pytorch

### ISSUE THREAD

**Issue by mrektor** [cta-observatory/dl1-data-handler#105]
When using a pytorch Dataloader you have the option of using multiple workers for reading the data (which vastly enhances the performances because it allow to read in parallel the data and have a 100% utilisation of the GPU).

With the current implementation of dl1dh this is not possible but there is a fix which is exposed here: https://github.com/pytorch/pytorch/issues/11929

The issue is relatively easy to solve
![image](https://user-images.githubusercontent.com/26904650/131825729-23a57b98-1b79-456d-b22c-1508ba20be37.png)

and the explanation is the following:
![image](https://user-images.githubusercontent.com/26904650/131825778-2220c8a5-3d48-4004-969c-4763494ba1c6.png)

**Comment by TjarkMiener** [comment]
Thanks for opening the issue and also providing a possible solution @mrektor! It would be great if the GammaLeaners @vuillaut @mikael10j can add their input.

**Comment by mikael10j** [comment]
Hi guys !
In principle, if you have enough memory to store all your data, it's ok to load them in the __init__ of your dataloader:
```
def __init__(self, file_path):
    with tables.open_file(file_path, 'r') as f:
        self.images = f.root[dl1_images_lstcam_key].col('image')
```
If you need to load your data from the file at each epoch because of memory limitations, the solution exposed by @mrektor is the one to follow. If you need to access the data (for filtering for instance) before the train process has started, you have to close the h5 file afterwards. For a complete example, you can take a look at the dataset classes of GammaLearn (https://gitlab.lapp.in2p3.fr/GammaLearn/GammaLearn/-/blob/master/gammalearn/datasets.py)

**Comment by maxnoe** [comment]
@mikael10j please not that newer versions of ctapipe support transformations of the image / peaktime columns to reduce storage size by using fixed precision.

You should use `ctapipe.io.read_table` to read the image table, which will invert those transformations transparently for the users.

This also connects to this issue: #104 (The title of that issue is not well chosen I'd say, it's about more efficiently loading images, by using chunked / whole table reading not reading single images from the hdf5 files).

**Comment by TjarkMiener** [comment]
Thanks for the quick input @mikael10j!

What the solution of @mrektor is suggesting is to revert [PR 88](https://github.com/cta-observatory/dl1-data-handler/pull/88), which turned out to be extremely useful for tensorflow data loading. I'd propose that we should crunch the numbers for pytorch and see if it's necessary to support this feature.

@mrektor Have you tried to increase the batch size so you exploit the whole capability of your GPU? That should have the same effect as parallel training on a single GPU. If you want to train a model with multiple GPUs, you may need that suggested feature. 

@maxnoe I'd say that issue #104 is entitled correctly and a fundamental discussion about switching back to chunked / whole table reading, as it was implemented when this effort started, deserved a separate issue.

**Comment by mrektor** [comment]
@TjarkMiener Yes I have changed the batch size but the issue is that the bottleneck becomes the reading of data instead of the computation. I have temporarily avoided the problem by loading a subset of the data in the ram of a big machine... but this is not scalable and the possibility of having this fix would be very important.

Is it a big problem to open the hdf5 at the first iteration rather than in the `__init__`?

**Comment by TjarkMiener** [comment]
hi again GammaLeaners @mikael10j @vuillaut
Ettore is working with files that are very similar to the one we used in the joint [hex preprocessing work at the ICRC2019](https://pos.sissa.it/358/753/pdf). Which version of GammaLearn was used for this work? The `pytorch` `DataLoader` of this version should be fully functional out of the box. I don't think the wheel have to be reinvented here. Can you please help us setting up the correct version? This would also enable a MAGIC analysis with GammaLearn.

@mrektor I quickly check the GammaLearn repo and my bet would be to use the `DataLoader` of the tree v0.4.X of GammaLearn, which can be found here: [https://gitlab.lapp.in2p3.fr/GammaLearn/GammaLearn/-/blob/065b021e65dde6717f04e8a7ab4833ac500ef395/gammalearn/datasets.py](https://gitlab.lapp.in2p3.fr/GammaLearn/GammaLearn/-/blob/065b021e65dde6717f04e8a7ab4833ac500ef395/gammalearn/datasets.py).

**Comment by mikael10j** [comment]
Indeed, `BaseDL1DHDataset`, `MemoryDL1DHDataset` and `FileDL1DHDataset` in GammaLearn 0.4 should be a good point to start. Then, dataset and dataloader creation is handled in `data_handlers.py`.
I could not find though the exact version we've used for our paper.

**Comment by TjarkMiener** [comment]
Should be solved by @mikael10j suggestion. I'm closing this, because I don't think it's a dl1dh issue, but feel free to reopen.
